File: src/main/scala/temportalist/compression/main/common/item/ICompressed.scala
package temportalist.compression.main.common.item

import java.util

import net.minecraft.client.gui.GuiScreen
import net.minecraft.entity.Entity
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.{Item, ItemStack}
import net.minecraft.world.World
import net.minecraftforge.client.settings.KeyModifier
import net.minecraftforge.fml.relauncher.{Side, SideOnly}
import temportalist.compression.main.common.Effects
import temportalist.compression.main.common.entity.EntityItemCompressed
import temportalist.compression.main.common.init.Compressed
import temportalist.origin.api.common.lib.Vect

/**
  *
  * Created by TheTemportalist on 4/14/2016.
  *
  * @author TheTemportalist
  */
trait ICompressed extends Item {

	override def getItemStackDisplayName(itemStack: ItemStack): String = {
		if (itemStack.hasTagCompound) Compressed.getDisplayName(itemStack)
		else super.getItemStackDisplayName(itemStack)
	}

	@SideOnly(Side.CLIENT)
	override def addInformation(stack: ItemStack, playerIn: EntityPlayer, tooltip: util.List[String],
			advanced: Boolean): Unit = {
		super.addInformation(stack, playerIn, tooltip, advanced)

		if (GuiScreen.isShiftKeyDown) {
			tooltip.add(Compressed.getTier(stack).toString)
			tooltip.add(Compressed.getSize(stack).toString)
			tooltip.add(this.getClass.getSimpleName)
			tooltip.add(Compressed.getStackName(stack))
		}

	}

	override def onUpdate(stack: ItemStack, worldIn: World, entityIn: Entity,
			itemSlot: Int, isSelected: Boolean): Unit = {
		entityIn match {
			case player: EntityPlayer =>
				if (!player.isSneaking)
					Effects.onInvUpdateCompressed(worldIn, player, stack)
			case _ =>
		}
	}

	override def hasCustomEntity(itemStack: ItemStack): Boolean = true

	override def createEntity(world: World, oldEI: Entity, itemStack: ItemStack): Entity = {
		new EntityItemCompressed(world, new Vect(oldEI),
			new Vect(oldEI.motionX, oldEI.motionY, oldEI.motionZ), itemStack)
	}



}

File: src/main/java/com/temportalist/compression/common/items/ICompressed.java
package com.temportalist.compression.common.items;

import com.temportalist.compression.common.effects.Effects;
import com.temportalist.compression.common.entity.EntityCompressed;
import com.temportalist.compression.common.init.CompressedStack;
import net.minecraft.client.gui.GuiScreen;
import net.minecraft.entity.Entity;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;

import java.util.List;

public interface ICompressed {

    default String getDisplayName(ItemStack stack) {
        if (stack.hasTagCompound()) return CompressedStack.getDisplayNameFor(stack);
        else return null;
    }

    default void addInfo(ItemStack stack, List<String> tooltip) {
        if (GuiScreen.isShiftKeyDown()) {
            tooltip.add(CompressedStack.getStackName(stack));
            tooltip.add(Long.toString(CompressedStack.getTier(stack).getSizeMax()));
        }
    }

    default boolean hasEntityCompressed() {
        return true;
    }

    default Entity createEntityCompressed(World world, Entity entity, ItemStack stack) {
        return new EntityCompressed(world,
                entity.getPositionVector(),
                new Vec3d(entity.motionX, entity.motionY, entity.motionZ),
                stack
        );
    }

    default void onUpdate(World world, Entity entity, ItemStack stack) {
        Effects.onUpdateEntityInventory(world, entity, stack);
    }

}
